hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java

```diff
@@ -726,13 +735,19 @@ public void afterObtainConnection(Connection connection) throws SQLException {
 			initialSchema = connection.getSchema();
 			connection.setSchema( tenantSchema() );
 		}
+		if ( readOnly ) {
+			connection.setReadOnly( true );
```

Comment:
IMO this stuff should happen at the `ConnectionProvider` level since invoking this method might not be so cheap as you can see here: https://github.com/pgjdbc/pgjdbc/issues/1228

Reply:
Also, that would allow `ConnectionProvider` implementations that serve connections from two dedicated pools, further reducing the cost.

Reply:
Hrrm, well, I *really* want to keep this together with `connection.setSchema()` since they're conceptually very similar. Originally had that somewhere else (forget where) but Steve wanted it to be called from the `LogicalConnection` so I moved it there.

Note that there's no chance of a performance _regression_ here, since `readOnly` will always be false for existing programs.

But I guess I need to dig a bit deeper on what precisely are the implications for Postgres.

Reply:
That looks good, but why not do this `handlesConnectionReadOnly` also for `org.hibernate.engine.jdbc.connections.spi.ConnectionProvider`? A custom connection provider could then decide if they want to optimize the read-only case.

Reply:
Yeah, I think that's reasonable.